pkg/handlers/v1/cloud_fetch.go

```diff
@@ -117,44 +164,113 @@ func (h *CloudFetchByHostnameHandler) Handle(ctx context.Context, input CloudAss
 	return extractOutput(assets), nil
 }
 
+func validateAssetType(input string) (string, error) {
+	switch input {
+	case awsEC2, awsELB, awsALB:
+		return input, nil
+	default:
+		return "", fmt.Errorf("unknown asset type %s", input)
+	}
+}
+
 // CloudFetchAllAssetsByTimeHandler defines a lambda handler for bulk fetching cloud assets known at specific point in time
 type CloudFetchAllAssetsByTimeHandler struct {
 	LogFn   domain.LogFn
 	StatFn  domain.StatFn
 	Fetcher domain.CloudAllAssetsByTimeFetcher
 }
 
-// Handle handles fetching cloud assets by hostname
-func (h *CloudFetchAllAssetsByTimeHandler) Handle(ctx context.Context, input CloudAssetFetchAllByTimestampParameters) (CloudAssets, error) {
+// Handle handles fetching cloud assets with pagination
+func (h *CloudFetchAllAssetsByTimeHandler) Handle(ctx context.Context, input CloudAssetFetchAllByTimestampParameters) (PagedCloudAssets, error) {
 	logger := h.LogFn(ctx)
 
 	ts, e := time.Parse(time.RFC3339Nano, input.Timestamp)
 	if e != nil {
 		logger.Info(logs.InvalidInput{Reason: e.Error()})
-		return CloudAssets{}, InvalidInput{Field: "time", Cause: e}
+		return PagedCloudAssets{}, InvalidInput{Field: "time", Cause: e}
 	}
 
-	if input.Count == nil {
+	if input.Count == 0 {
 		e = errors.New("missing or malformed required parameter count")
 		logger.Info(logs.InvalidInput{Reason: e.Error()})
-		return CloudAssets{}, InvalidInput{Field: "count", Cause: e}
+		return PagedCloudAssets{}, InvalidInput{Field: "count", Cause: e}
 	}
 
-	var offset uint
-	if input.Offset != nil {
-		offset = *input.Offset
+	assetType, e := validateAssetType(input.Type)
+	if e != nil {
+		logger.Info(logs.InvalidInput{Reason: e.Error()})
+		return PagedCloudAssets{}, InvalidInput{Field: "type", Cause: e}
 	}
 
-	assets, e := h.Fetcher.FetchAll(ctx, ts, *input.Count, offset)
+	var offset uint // do not offset as this is the first page
+	assets, e := h.Fetcher.FetchAll(ctx, ts, input.Count, offset, assetType)
 	if e != nil {
 		logger.Error(logs.StorageError{Reason: e.Error()})
-		return CloudAssets{}, e
+		return PagedCloudAssets{}, e
 	}
 	if len(assets) == 0 {
-		return CloudAssets{}, NotFound{ID: "any"}
+		return PagedCloudAssets{}, NotFound{ID: "any"}
 	}
 
-	return extractOutput(assets), nil
+	nextPageToken, e := input.toNextPageToken()
+	if e != nil {
+		logger.Error(logs.StorageError{Reason: e.Error()})
```

Comment:
question: Why do we continue here instead of returning the storage error?

Reply:
The current page of results is valid, and the case where we can not generate the token from valid parameters is unlikely (if at all possible), but I did not want to ignore the possible error condition that we did not foresee. So the logic currently is - return valid results (that's what client requested), log any issue with generating token for the next page, and let the next request fail.
One clear action item - I should not use StorageError for this, as it has nothing to do with storage, but I am open to failing loudly if we can't  generate token for the next page even when we have valid results for the current one.